### PR TITLE
Adding Derived type for parameter dependency

### PR DESCRIPTION
**Pull request by aramirezreyes**
In this branch I will give it a go to implemented the feature described in #358. I have never programmed a macro so I will try to follow #169.  Advice from @sebastianpech is welcome!

### CODE REVIEW

**Comment by aramirezreyes** [comment]
This is a prototype that I thought of today. To date it does not break any test. 
So far, a ComputedParameter can only depend on 1 pre-existing parameter.

Does this design make sense @Datseris?

**Comment by Datseris** [comment]
This does make sense to me! But also, shouldn't we allow the produced parameter to depend in as many parameters as the user wants...?

Just make the field of the parameter a vector of values instead, and if a user gives a single string/symbol make it into a vector anyways.

**Comment by aramirezreyes** [comment]
I added this change and some docs. Now the use cases I added to the tests are:

```julia

p = Dict(:α => [1, 2],
    :solver => [SolverA,SolverB],
    :β => ComputedParameter(:α, x -> x^2),
    )


@test Set([ Dict(:α => 1, :solver => SolverA, :β => 1),
    Dict(:α => 2, :solver => SolverA, :β => 4),
    Dict(:α => 1, :solver => SolverB, :β => 1),
    Dict(:α => 2, :solver => SolverB, :β => 4),
    ]) == Set(dict_list(p))

p2 = Dict(:α => [1, 2],
    :β => [10,100],
    :solver => [SolverA,SolverB],
    :γ => ComputedParameter([:α,:β], (x,y) -> x^2 + 2y),
    )

@test Set([ Dict(:α => 1, :solver => SolverA, :β => 10, :γ => 21),
    Dict(:α => 2, :solver => SolverA, :β => 10, :γ => 24),
    Dict(:α => 1, :solver => SolverB, :β => 10, :γ => 21),
    Dict(:α => 2, :solver => SolverB, :β => 10, :γ => 24),
    Dict(:α => 1, :solver => SolverA, :β => 100, :γ => 201),
    Dict(:α => 2, :solver => SolverA, :β => 100, :γ => 204),
    Dict(:α => 1, :solver => SolverB, :β => 100, :γ => 201),
    Dict(:α => 2, :solver => SolverB, :β => 100, :γ => 204),
   ]) == Set(dict_list(p2))
```

**Comment by Datseris** [comment]
This is great! You should
0. Consider renaming the whole `ComputedParameter` to just `Derived`. Everything is computed after all :) 
1. Increase minor version in Project.toml
2. Add a changelog entry for this
3. Add a refenrece to `Derived` in the docstring of `dict_list`.
4. Actually add `Derived` to the documentaiton, which means go to the file `run&list.md`, and in the `@docs` block that has `dict_list`, also add `Derived`.

then I'll merge and tag!

File: src/dict_list.jl [hunk 1]
```diff
@@ -275,3 +275,82 @@ julia> dict_list(d) # only in case `:a` is `1` the dictionary will get key `:c`
...
+
+"""
+    ComputedParameter(independentP <: Union{String,Symbol}, func :: Function)
+Constructs a ComputedParameter from a single independent parameter.
+"""
+function ComputedParameter(independentP :: Union{String,Symbol}, func :: Function) 
+    return ComputedParameter([independentP], func)
+end
+
+export ComputedParameter
```
Comment (Datseris) [inline src/dict_list.jl:340]: Also please move this to the top of the file with the other export statements

File: src/dict_list.jl [hunk 2]
```diff
@@ -275,3 +275,82 @@ julia> dict_list(d) # only in case `:a` is `1` the dictionary will get key `:c`
...
+
+"""
+   produce_computed_parameter(dicts)
+Receive an array of parameter dictionaries, and for each one, evaluates the computed parameters after the possible combination of parameters has been created.
+"""
+function produce_computed_parameters(dicts)
+    map(dicts) do dict
```
Comment (Datseris) [inline src/dict_list.jl:347]: By the way, we shouldn't `map` here. This creates a container that we don't need. We should simple loop over the `dict` in `dicts` and perform the subsequent code!

File: src/dict_list.jl (hunk 1 above)
Comment (aramirezreyes) [reply src/dict_list.jl]: Oh yes. Actually, this file did not have the exports at the top. I took the liberty of moving them up.

File: src/dict_list.jl (hunk 2 above)
Comment (aramirezreyes) [reply src/dict_list.jl]: Fixed!

**Comment by aramirezreyes** [comment]
I think everything is addressed!

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,6 @@
+# 2.11.0
+* Add `Derived` type and function to allow `dict_list` to compute some parameters from the value of other parameters.
+* Moved `export` statements to the top of the file in `dict_list.jl`
```
Comment (Datseris) [inline CHANGELOG.md:3]:
```suggestion

```
This has no reason to be mentioned in the changelog xD

File: docs/src/real_world.md [hunk 4]
```diff
@@ -279,6 +279,24 @@ dicts = dict_list(general_args)
 println("Total dictionaries made: ", length(dicts))
 dicts[1]
 ```
+Also, using the type [`ComputedParameter`](@ref), we can have parameters that are computed depending on the value of other parameters:
```
Comment (Datseris) [inline docs/src/real_world.md:282]:
```suggestion
Also, using the type [`Derived`](@ref), we can have parameters that are computed depending on the value of other parameters:
```

File: src/dict_list.jl [hunk 5]
```diff
@@ -275,3 +298,85 @@ julia> dict_list(d) # only in case `:a` is `1` the dictionary will get key `:c`
...
+struct Derived{T}
+    independentParam::Vector{T}
+    func::Function
+end
+
+"""
+    Derived(independentP <: Union{String,Symbol}, func :: Function)
+Constructs a Derived from a single independent parameter.
+"""
+function Derived(independentP :: Union{String,Symbol}, func :: Function)
```
Comment (Datseris) [inline src/dict_list.jl:362]: weird syntax to have white space between `::`. I/ve never seen it before. Should be without white space.

File: src/dict_list.jl [hunk 6]
```diff
@@ -275,3 +298,85 @@ julia> dict_list(d) # only in case `:a` is `1` the dictionary will get key `:c`
      end
      :(toDependentParameter($(esc(value)),$(esc(condition))))
  end
+
+"""
+   Derived(parameters::Vector{Union{String,Symbol}},function::Function)
```
Comment (Datseris) [inline src/dict_list.jl:303]:
```suggestion
   Derived(parameters::Vector{Union{String,Symbol}}, function::Function)
```

File: src/dict_list.jl [hunk 7]
```diff
@@ -275,3 +298,85 @@ julia> dict_list(d) # only in case `:a` is `1` the dictionary will get key `:c`
...
+Wrap the name(s) of a parameter(s) and a function. After the 
+possible parameter combinations are created, dict_list will replace instances of
```
Comment (Datseris) [inline src/dict_list.jl:305]:
```suggestion
possible parameter combinations are created, [`dict_list`](@ref) will replace instances of 
```

File: src/dict_list.jl [hunk 8]
```diff
@@ -275,3 +298,85 @@ julia> dict_list(d) # only in case `:a` is `1` the dictionary will get key `:c`
...
+Constructs a Derived from a single independent parameter.
+"""
+function Derived(independentP::Union{String,Symbol}, func::Function) 
+    return Derived([independentP], func)
+end
+
+
+"""
+   produce_computed_parameter(dicts)
+Receive an array of parameter dictionaries, and for each one, evaluates the computed parameters after the possible combination of parameters has been created.
```
Comment (Datseris) [inline src/dict_list.jl:369]:
```suggestion
Receive an array of parameter dictionaries, and for each one, evaluate
the computed parameters after the possible combination of 
parameters has been created.
```

File: docs/src/real_world.md [hunk 9]
```diff
@@ -279,6 +279,24 @@ dicts = dict_list(general_args)
...
+```@example customizing
+using DrWatson
+
+general_args2 = Dict(
+    "model" => "barkley",
+    "noise" => [0.075, 0.050, 0.025],
+    "noise2" => [1.0, ComputedParameter(["noise", "N"], (x,y) -> 2x + y)],
+    "noisy_training" => true,
+    "N" => 100,
+       )
```
Comment (Datseris) [inline docs/src/real_world.md:292]:
```suggestion
)
```

**Comment by Datseris** [comment]
thanks, I Think that's a great feature :)

**Comment by aramirezreyes** [comment]
Thanks!

**Pull request merged**